File: LeetCode/Python/MinimumNumberOfPeopleToTeach.py
from typing import List
class Solution:
    def minimumTeachings(self, n: int, languages: List[List[int]], friendships: List[List[int]]) -> int:
        need = set()
        for u, v in friendships:
            u -= 1
            v -= 1
            ok = False
            for x in languages[u]:
                if x in languages[v]:
                    ok = True
                    break
            if not ok:
                need.add(u)
                need.add(v)

        ans = len(languages) + 1
        for i in range(1, n + 1):
            cans = 0
            for v in need:
                if i not in languages[v]:
                    cans += 1
            ans = min(ans, cans)
        return ans
    
    
s=Solution()

n = 2
languages = [[1],[2],[1,2]]
friendships = [[1,2],[1,3],[2,3]]
print("Minimum Teachings : ",s.minimumTeachings(n,languages,friendships))

File: LeetCode/Java/MinimumNumberOfPeopleToTeach.java
import java.util.*;

public class MinimumNumberOfPeopleToTeach {

    public int minimumTeachings(int n, List<List<Integer>> languages, int[][] friendships) {
        Set<Integer> need = new HashSet<>();

        for (int[] f : friendships) {
            int u = f[0] - 1;
            int v = f[1] - 1;
            boolean ok = false;

            for (int langU : languages.get(u)) {
                if (languages.get(v).contains(langU)) {
                    ok = true;
                    break;
                }
            }
            if (!ok) {
                need.add(u);
                need.add(v);
            }
        }

        int ans = languages.size() + 1;

        for (int i = 1; i <= n; i++) {
            int cans = 0;
            for (int v : need) {
                if (!languages.get(v).contains(i)) {
                    cans++;
                }
            }
            ans = Math.min(ans, cans);
        }

        return ans;
    }

    public static void main(String[] args) {
        MinimumNumberOfPeopleToTeach s = new MinimumNumberOfPeopleToTeach();

        int n = 2;
        List<List<Integer>> languages = new ArrayList<>();
        languages.add(Arrays.asList(1));
        languages.add(Arrays.asList(2));
        languages.add(Arrays.asList(1, 2));

        int[][] friendships = {{1, 2}, {1, 3}, {2, 3}};

        System.out.println("Minimum Teachings : " + 
            s.minimumTeachings(n, languages, friendships));
    }
}
